articles/remapping.md

```diff
@@ -0,0 +1,119 @@
+It must be possible to make a rule that replaces "big/red" with "small" such that the final names are `/small/dog`, `/very_big/red/dog`
+
+##### 2. Basename Replacement
+**A user can replace a full basename token with another basename token**
+
+*Example*
+Node uses names `/big/cat`, `/big/cat/paws`.
+It must be possible to make a rule that replaces "cat" with "dog" such that the final names are `/big/dog`, `/big/cat/paws`
+
+##### 3. Exact Replacement
+**A user can replace a full name with another name**
+
+*Example*
+Node uses names `/big/cat`, `/big/cat/paws`.
+It must be possible to make a rule that replaces "/big/cat" with "/bear" such that the final names are `/bear`, `/big/cat/paws`
+
+##### 4. Arbitrary Substring Replacement ***(maybe)***
+**A user can replace any substring with another string**
+
+*Examples*
+Node uses names `/big/red/dog`, `/very_big/red/dog`.
+It must be possible to make a rule that replaces "big/red" with "small" such that the final names are `/small/dog`, `/very_small/dog`
+
+Node uses names `/big/dog`, `/dog/big`.
+It must be possible to make a rule that replaces "dog" with "cat" such that the final names are `/big/cat`, `/cat/big`
+
+Node uses names `/big/red/dog`, `/small/cat`.
+Node is run in namespace `ns`.
+It must be possible to make a rule that replaces "/" with "_" such that the final names are `/ns/_big_red_dog`, `/ns/_small_cat`
+
+#### Static Remapping
+Static remapping is the ability to provide a node with remap rules prior to launching it.
+A node stores all remapping rules for the duration of it's life.
+All names have remapping rules applied before they are used.
+This means the DDS interface, and maybe all of `rmw`, is ignorant of remapping.
+
+In ROS 1 remap rules are passed into the node at startup.
+`roscpp` received remap rules as arguments to `ros::init()`.
+`rospy` loaded remap rules from `sys.argv` during `rospy.init_node()`.
+
+In ROS 2 this code can be shared between client implementations by doing it in `rcl`.
+[rcl_init()](http://docs.ros2.org/beta1/api/rcl/rcl_8h.html#a6abc5188c50b3a4a5fc41b357b49d6e5) already requires command line arguments to be passed to it.
+It could parse the rules into an intermediate representation and store it internally.
+Methods like [rcl_publisher_init()](http://docs.ros2.org/beta1/api/rcl/publisher_8h.html#a0f09bd85795259f6c96f38c875d19d13) and [rcl_subscription_init()](http://docs.ros2.org/beta1/api/rcl/subscription_8h.html#a2b8a58f9ae9fef8adba35b5a45db10a0) would apply remapping rules automatically.
+Making name remapping an automatic part of `rcl` suggests that features using names, such as parameters, would be implemented in `rcl` and not `rclcpp` and `rclpy` as indicated on [slide 31](http://roscon.ros.org/2016/presentations/ROSCon%202016%20-%20ROS%202%20Update.pdf).
+
+#### Dynamic Remapping
+
+Dynamic remapping is the ability to remap a name after it has already been used.
+For a neat user experience, Dynamic remapping requires rules to be chainable.
```

Comment:
I think this needs more explanation. While we have talked about it a lot and know why it would be good to provide more information to readers why that is "neat" / "sane".